r/src/compute.cpp

```diff
@@ -223,4 +205,66 @@ std::shared_ptr<arrow::Table> Table__FilterChunked(
   }
   return tab;
 }
+
+arrow::Datum to_datum(SEXP x) {
+  // TODO: this is repetitive, can we DRY it out?
+  if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, "Array")) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<arrow::Array>> obj(x);
+    return static_cast<std::shared_ptr<arrow::Array>>(obj);
+  } else if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, "ChunkedArray")) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<arrow::ChunkedArray>> obj(x);
+    return static_cast<std::shared_ptr<arrow::ChunkedArray>>(obj);
+  } else if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, "RecordBatch")) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<arrow::RecordBatch>> obj(x);
+    return static_cast<std::shared_ptr<arrow::RecordBatch>>(obj);
+  } else if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, "Table")) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<arrow::Table>> obj(x);
+    return static_cast<std::shared_ptr<arrow::Table>>(obj);
+  } else {
+    // TODO: scalar?
+    // This assumes that R objects have already been converted to Arrow objects;
+    // that seems right but should we do the wrapping here too/instead?
+    Rcpp::stop("to_datum: Not implemented");
+  }
+}
+
+SEXP from_datum(arrow::Datum datum) {
+  if (datum.is_array()) {
+    return Rcpp::wrap(datum.make_array());
+  } else if (datum.is_arraylike()) {
+    return Rcpp::wrap(datum.chunked_array());
+  } else {
+    // TODO: the other datum types
+    Rcpp::stop("from_datum: Not implemented");
+  }
+}
+
+std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(std::string func_name,
+    List_ options) {
+  if (func_name == "filter") {
+    auto out = std::make_shared<arrow::compute::FilterOptions>(arrow::compute::FilterOptions::Defaults());
+    if (!Rf_isNull(options["keep_na"]) && options["keep_na"]) {
+      out->null_selection_behavior = arrow::compute::FilterOptions::EMIT_NULL;
+    }
+    return out;
+  } else if (func_name == "take") {
+    auto out = std::make_shared<arrow::compute::TakeOptions>(arrow::compute::TakeOptions::Defaults());
+    return out;
+  } else {
+    return nullptr;
+  }
+  // TODO: make sure the correct destructor is called?
```

Comment:
This was @bkietz's note to self, I think

Reply:
```suggestion
```
confirmed: `make_shared<T>` sets the destructor to `deleter<T>` (so we aren't relying on `~FunctionOptions` being virtual)